File: scripts/embed.ts
import { PGEssay, PGJSON } from "@/types";
import { loadEnvConfig } from "@next/env";
import { createClient } from "@supabase/supabase-js";
import fs from "fs";
import { Configuration, OpenAIApi } from "openai";

loadEnvConfig("");
const apikey = process.env.AZURE_OPENAI_APIKEY;
const baseurl = process.env.AZURE_OPENAI_ENDPOINT;
const deploymentname = process.env.AZURE_OPENAI_DEPLOYMENT;
let base_url = `${baseurl}openai/deployments/${deploymentname}`;
//let url = `${baseurl}openai/deployments/${deploymentname}/embeddings?api-version=2022-12-01`;

const generateEmbeddings = async (essays: PGEssay[]) => {
  //const configuration = new Configuration({ apiKey: process.env.OPENAI_API_KEY });
  //const openai = new OpenAIApi(configuration);
  const configuration = new Configuration({
    basePath: base_url,
    apiKey: apikey,
  });
  const openai = new OpenAIApi(configuration);

  const supabase = createClient(process.env.NEXT_PUBLIC_SUPABASE_URL!, process.env.SUPABASE_SERVICE_ROLE_KEY!);

  for (let i = 0; i < essays.length; i++) {
    const section = essays[i];

    for (let j = 0; j < section.chunks.length; j++) {
      const chunk = section.chunks[j];

      const { essay_title, essay_url, essay_date, essay_thanks, content, content_length, content_tokens } = chunk;

      const embeddingResponse = await openai.createEmbedding({
        deployment: "text-embedding-ada-002",
        input: content
      },
      {
        headers: {
          "api-key": apikey,
        },
        params: {
          "api-version": "2022-12-01",
        },
      });

      const [{ embedding }] = embeddingResponse.data.data;

      const { data, error } = await supabase
        .from("pg")
        .insert({
          essay_title,
          essay_url,
          essay_date,
          essay_thanks,
          content,
          content_length,
          content_tokens,
          embedding
        })
        .select("*");

      if (error) {
        console.log("error", error);
      } else {
        console.log("saved", i, j);
      }

      await new Promise((resolve) => setTimeout(resolve, 800));
    }
  }
};

(async () => {
  const book: PGJSON = JSON.parse(fs.readFileSync("scripts/pg.json", "utf8"));

  await generateEmbeddings(book.essays);
})();

File: scripts/embed.py
import json
import time
from enum import Enum

from openai import Configuration, OpenAiApi
from supabase_py import create_client
from dotenv import load_dotenv
import os


load_dotenv()  # loading environment variables from .env file


# Enum for constants
class Constants(Enum):
    OPENAI_API_KEY = os.getenv("AZURE_OPENAI_APIKEY")
    OPENAI_ENDPOINT = os.getenv("AZURE_OPENAI_ENDPOINT")
    OPENAI_EMBEDDING = os.getenv("AZURE_OPENAI_EMBEDDING")
    OPENAI_VERSION = os.getenv("AZURE_OPENAI_VERSION")


OPENAI_BASE_URL = f"{Constants.OPENAI_ENDPOINT.value}openai/deployments/{Constants.OPENAI_EMBEDDING.value}"


# Generate embeddings for essays
def generate_embeddings(essays):
    configuration = Configuration(
        base_path=OPENAI_BASE_URL,
        api_key=Constants.OPENAI_API_KEY.value
    )
    openai = OpenAiApi(configuration=configuration)

    supabase_url = os.getenv("NEXT_PUBLIC_SUPABASE_URL")
    supabase_key = os.getenv("SUPABASE_SERVICE_ROLE_KEY")
    supabase = create_client(supabase_url, supabase_key)

    for i, section in enumerate(essays):
        for j, chunk in enumerate(section["chunks"]):
            essay_title = chunk["essay_title"]
            essay_url = chunk["essay_url"]
            essay_date = chunk["essay_date"]
            essay_thanks = chunk["essay_thanks"]
            content = chunk["content"]
            content_length = chunk["content_length"]
            content_tokens = chunk["content_tokens"]

            # use try-except block for error handling
            try:
                embedding_response = openai.create_embedding(
                    model=Constants.OPENAI_EMBEDDING.value,
                    input_=content,
                    options={
                        "api_key": Constants.OPENAI_API_KEY.value,
                        "parameters": {
                            "api_version": Constants.OPENAI_VERSION.value
                        }
                    }
                )
                
                #extracting the embeddings
                embedding = embedding_response["data"]["data"][0]["embedding"]
                
                # insert data and handling supabase errors
                data, error = supabase.table("pg").insert({
                    "essay_title": essay_title,
                    "essay_url": essay_url,
                    "essay_date": essay_date,
                    "essay_thanks": essay_thanks,
                    "content": content,
                    "content_length": content_length,
                    "content_tokens": content_tokens,
                    "embedding": embedding
                }).execute()

                if error:
                    print("error", error)
                else:
                    print(f"saved {i}, {j} {essay_title} {content_tokens}")

                time.sleep(0.8)
            
            except Exception as e:
                print(f"error processing essay {i}, {j}, {e}")
    

if __name__ == '__main__':
    with open('scripts/pg.json', 'r') as f:
        book = json.load(f)

    generate_embeddings(book["essays"])
